### DISCUSSION TITLE
Paged.js example

### DISCUSSION THREAD

**Discussion by kbvernon** [quarto-dev/quarto-cli#6401] [Show and Tell]
### Description

I don't know if this is helpful, but while people are waiting for Quarto support for Paged.js, I put together this example showing how to do it manually:

Site: [kbvernon.github.io/quarto-pagedjs/](https://kbvernon.github.io/quarto-pagedjs/)
Repo: [https://github.com/kbvernon/quarto-pagedjs](https://github.com/kbvernon/quarto-pagedjs)

I'm probably doing something wrong, but it works for my needs. Have to generate a bunch of pre-filled forms soon, and, well, I hate LaTeX. I did some experiments with Typst, and I can confirm that it's way way way better than LaTeX (that's not saying much), but I already know HTML/CSS, so getting Paged.js working was way less effort.

Anyway, I'll stop rambling. As always, thanks for being awesome!

**Comment by jimjam-slam** [comment]
I'm 100% with you, @kbvernon! There's obviously still work to do to make Paged PDFs a first-class Quarto citizen, but you can get pretty far already with some basic styles.

If you want to compare notes, here's some of what I did to run off collections of Quarto docs as a PDF via Paged.js (or weasyprint, whch I tried first—but they both work pretty similarly): https://github.com/quarto-dev/quarto-cli/discussions/2538#discussioncomment-3739791

I'll be sure to check out the source on yours too!

**Reply by kbvernon** [reply]
Ha. Thanks, @jimjam-slam! I think I ran across that issue/discussion awhile back. It's one of the few breadcrumbs that you find littered across the web when you - or, I should say "when I" - Google "Quarto Paged.js." I guess using `paged.polyfill.js` doesn't require a complete rebuild of default css, which is good. On the other hand, `pagedown::chrome_print()` is hanging on the rendered html file for some reason, though that could be specific to something I'm doing - can never rule out user error over here... At any rate, the result is that I have to open in Chrome and print to PDF manually, which is definitely not preferable.

**Reply by kbvernon** [reply]
Hi, @jimjam-slam, do you by chance have a repo where you try out Paged.js? I'd like to see how you did it. Thanks!

**Reply by jimjam-slam** [reply]
How's this, @kbvernon? https://github.com/jimjam-slam/quarto-experiments-pagedbook

The only problem I have is that I was previously able to retain the HTML, and I can no longer remember how to do that—I resorted to trying to copy and paste the intermediary HTML in the brief moment before it gets deleted (at the end of the render). The HTML is pretty barebones anyway (I ended up wrapping my chapters in a div to make targeting them easier)! But hopefully it's a start 😊

**Reply by jimjam-slam** [reply]
`pdf-engine: weasyprint` also works, but I do see some minor rendering differences, and the CSS support isn't quite as good IMO!

**Reply by kbvernon** [reply]
Ah! This is really helpful! Thanks for sharing! I was going a round-about way through Chrome to get to the pdf. This looks a lot easier. 

Can you still pass an html template this way?

**Reply by jimjam-slam** [reply]
Sorry, I never made an HTML template! I just did this with Quarto's default template 😅

**Reply by kbvernon** [reply]
Yeah, I'll have to experiment with it. First need to get pandoc to find pagedjs-cli.cmd... 

But, anyway, thanks for your help!

**Reply by jimjam-slam** [reply]
Are you trying to run pandoc directly? Generally I found that if `pagedjs-cli` was on the path, `quarto render` worked fine!

Happy to help 😊

**Reply by kbvernon** [reply]
Yeah, I added pagedjs-cli.cmd to the path, but pandoc still can't find it. Trying to specify the full path in `pdf-engine: ...` doesn't work, either. Looks like there's an issue with pandoc + pagedjs on Windows machines: https://github.com/jgm/pandoc/issues/8061.

**Reply by jimjam-slam** [reply]
Mmm, that's frustrating. I'm on a Mac, so I hadn't noticed. If you're feeling gusty, you could run a GitHub Actions workflow to render it on a Linux machine!

**Reply by kbvernon** [reply]
~~Hm... your approach using `pdf-engine: pagedjs-cli` appears to be working for me now. I think the fix might have been as simple as restarting my machine after adding to the path on Windows. I guess I was looking for a bug somewhere that wasn't just user error...~~

Scratch that. Still very buggy on Windows.

**Comment by Bosmeneer** [comment]
Hi @kbvernon 
I'm trying to use your code, outputting HTML works. But PDF does not. What should I change in the qmd file to also output a PDF file when running it?

Many thanks

**Reply by mcanouil** [reply]
@TheBossmann Unfortunately "But PDF does not." is not enough for anyone to help you.

Adding `pdf: default` under format is enough to render.

If you don't want the R package for the lorem ipsum placeholder, then you can use the 1.5 built-in shortcode:

````qmd
---
title: "Paged.js"
format:
  pdf: default
  html:
    include-in-header: 
      - text: |
          <script src="paged.polyfill.js"></script>
      - text: |
          <link rel="stylesheet" type="text/css" href="print.css">
      - text: |
          <link rel="stylesheet" type="text/css" href="preview.css" media="screen">
    page-layout: custom
    embed-resources: true
---

Documentation for Paged.js here: <https://pagedjs.org/>. It's not terribly helpful, unfortunately.

In the yaml, we add the following as text fields to `include-in-header`:  

- `<script src="paged.polyfill.js"></script>`   
    This is the default source for Paged.js, which must be included in the header for any of this to work.  
- `<link rel="stylesheet" type="text/css" href="print.css">`  
    Paged.js uses this to control what happens to the layout when you click "Print" in your browser.  
- `<link rel="stylesheet" type="text/css" href="preview.css" media="screen">`  
    Gives the content a paged look when viewed in your browser. Note the `media="screen"` attribute! It's required!  

{{< lipsum 7 >}}
````

**Reply by kbvernon** [reply]
@mcanouil, just to clarify, including `pdf: default` would _not_ make the css formatting available to the pdf output, right? That would just generate a pdf with the default Quarto formatting.

@TheBossmann, if it helps, and assuming you use R, I have taken to outputting to html, then running `pagedown::chrome_print()` on the result. If you don't use R, you can always just manually print to pdf in your preferred browser. Only thing you would want to make sure to do is include [`@page`](https://developer.mozilla.org/en-US/docs/Web/CSS/@page) in your css file. This controls printing of a webpage.

**Reply by mcanouil** [reply]
LaTeX does not support HTML/CSS and there is roughly zero chance it ever will.

**Reply by kbvernon** [reply]
Thanks, I understand that. I guess my question was just really unclear. I was basically asking if there is now support for paged.js or similar, but I realize now even that was kind of a stupid question in this context...

**Reply by mcanouil** [reply]
Did you try Typst? (Rather than pdf/LaTeX)
It offers many things such as support for CSS in the upcoming Quarto release or the one after.

In order to be clear, Paged.js does not produce PDF. It produces a paged HTML that a user can print to a PDF (or directly to paper) using the "print" feature of any browser.

(I might at some point make a Quarto extension for Paged.js)

**Reply by kbvernon** [reply]
Typst/CSS sounds awesome! Of course, HTML/CSS is still preferable since the population of people who speak HTML is orders of magnitude larger than Typst (a dangerous argument for someone who lives and breathes R), but anyway having that Typst option would still be great. 

If you have plans to build a Paged.js extension, that would also be amazing. I would just like to have a `format: paged` or similar that allows you to render a qmd and then the process of compiling to HTML and printing to PDF is all done in one go (basically, `{pagedown}`). So, if there's any way I can help with that, just let me know!

**Reply by mcanouil** [reply]
> having that Typst option would still be great.

Would? It is there since the release of Quarto 1.4

> If you have plans to build a Paged.js extension, that would also be amazing. I would just like to have a format: paged or similar that allows you to render a qmd and then the process of compiling to HTML and printing to PDF is all done in one go (basically, {pagedown}). So, if there's any way I can help with that, just let me know!

Major constraint is time unfortunately.
